### DISCUSSION TITLE
Hydrostatic equilibrium boundary conditon

### DISCUSSION THREAD

**Discussion by kyriakosTapinou** [PrincetonUniversity/athena#623] [Q&A]
I've been trying to establish a hydrostatic equilibrium with constant gravity. I've used the Rayliegh-Taylor example as a starting point. Using the same style of pressure boundary condition where the ghost cell pressure is integrated from the domain pressure value is leading to pressure waves emitted on the boundary. Do you have any suggestions or resources you can point me to? 

In the current case it is all hydrodynamic and there are no fields. My pressure and density profile are not linear throughout the domain.

**Comment by tomidakn** [comment]
I guess the pressure waves are emitted because the boundary conditions are not sufficiently accurate and the ghost cells are (slightly) out of equilibrium. You can quantify it by measuring the pressure gradient force between the first active cell and first ghost cell. I suggest you to improve the accuracy of the extrapolation in the calculation of the ghost cell values, as the current one implemented in the RT pgen does not look very accurate.

Alternatively, you may directly apply boundary conditions on the flux. However, you need to modify the code by yourself as we do not provide such an interface at this moment. The fluxes from the Riemann solver are stored in the x(1/2/3)flux arrays, and you can modify them before they are applied in Hydro::AddFluxDivergence.

**Reply by kyriakosTapinou** [reply]
Thanks for the response - I'm trying to implement the following remedy now but I think it is something you would deem to specific to be introduced into the main branch. 

https://www.aanda.org/articles/aa/abs/2019/11/aa36387-19/aa36387-19.html

**Reply by kyriakosTapinou** [reply]
@tomidakn I've implemented this and get improved results for the case of a simple hydrostatic equilibirum. 

I'm combining this with data driving on the boundary and I need to access a generic meshblock's level within the ProblemGenerator function. I can't access the meshblock tree of the pmy_mesh as its private, is there another way to access meshblock characteristics (not prims or cons)? 

E.g. what I would like is to get the level of a meshblock (lx1, lx2, lx3) from inside ProblemGenerator()

**Reply by tomidakn** [reply]
You do not need to access the tree to get the location of a MeshBlock. MeshBlock::loc contains lx1,lx2,lx3 and level.

**Reply by kyriakosTapinou** [reply]
@tomidakn  sorry I wasn't clear. 

I have access to a specific meshblock with its own MeshBlock::loc, I need to find the "base" block e.g. if the meshblock I have access to currently with 
loc.lx1=3
loc.lx2 = 4
loc.lx3 = 3
I would need the level of the meshblock with location 
lx1=3
loc.lx2 = 4
lx3 = 0
which might even belong to a different process. I was hoping there was a global tree or some data structure with the level and location information.

**Comment by chengcli** [comment]
I dealt with this issue before. Instead of extending the pressure to the ghost zones, I use the existing reflective boundary condition but applied to "nonhydrostatic" pressure only. Here is my publication https://iopscience.iop.org/article/10.3847/1538-4365/aafdaa/meta on it and with the code available in this repo https://github.com/chengcli/canoe

**Reply by kyriakosTapinou** [reply]
@chengcli thank you I will have a look at this now!
